Products/ZenModel/MaintenanceWindow.py

```diff
@@ -653,14 +680,31 @@ def _setProdState(devices_batch):
         else:
             processFunc = _setProdState
 
+        # Adding exception handling for the following:
+        # ConflictError, POSKeyError and ReadConflictError.
+        #
+        # If any of the listed exceptions are encountered, we will retry
+        # devices in the batch individually.  If batchSize isn't
+        # specified and we encounter one of these exceptions, we will
+        # then specify a batch size and retry the devices in batches.
+        # If there are exceptions in these batches, we will retry the
+        # devices individually.  This will ensure that maintenance
+        # windows as a whole do not fail due to these exceptions.
+        # Fixes ZEN-31805.
         if batchSize:
-            for i in xrange(0, len(devices), batchSize):
-                log.info('MW %s processing batch #%s', self.displayName(),
-                         i / batchSize + 1)
-                processFunc(devices[i:i + batchSize])
+            processBatchOfDevices(devices, batchSize)
         else:
-            processFunc(devices)
+            try:
+                processFunc(devices)
+            except (ConflictError, POSKeyError, ReadConflictError) as e:
+                log.warn("Exception encountered and no batchSize specified. %s. Retrying in batches." % e)
```

Comment:
And here:
```
log.warn(
    "Exception encountered and no batchSize specified. "
    "%s. Retrying in batches.",
    e,
)
```
FYI, Python will automatically concatenate strings are separate lines when inside parenthesis.  Since the format string is argument to a function, that's considered to be inside parenthesis.

Reply:
fixed